### ISSUE TITLE
Run apply-all only on specified targets

### ISSUE THREAD

**Issue by robinbowes** [gruntwork-io/terragrunt#607]
Following on from #355, I'd like to be able to run terraform on a bunch of targets matching a pattern, ie. an "include" rather than an "exclude".

The use case is, given a tree structure like this:

```
├── siteA
│   ├── regionA
│   │   ├── service_A
│   │   ├── service_B
│   │   ├── service_C
│   │   ├── service_D
│   │   ├── service_E
│   │   ├── service_F
│   │   └── service_G
│   └── regionB
│       ├── service_A
│       ├── service_B
│       ├── service_C
│       ├── service_D
│       ├── service_E
│       ├── service_F
│       └── service_G
└── siteB
    ├── regionA
    │   ├── service_A
    │   ├── service_B
    │   ├── service_C
    │   ├── service_D
    │   ├── service_E
    │   ├── service_F
    │   └── service_G
    └── regionB
        ├── service_A
        ├── service_B
        ├── service_C
        ├── service_D
        ├── service_E
        ├── service_F
        └── service_G
```

I'd like to be able to run terraform on, eg. all service_C and service_D targets.

I actually have many more sites/regions/and services than shown above! :)

**Comment by brikis98** [comment]
Seems like a handy feature to have. PR welcome 👍

**Comment by robinbowes** [comment]
> PR welcome 👍

Bah!

**Comment by mattiarossi** [comment]
Hi,
I am working on a PR to provide this functionality, and have some questions regarding how to cope with running it in combination with the exclude-dirs flag
My take is that when running with include-dirs it makes no sense to add an exclusion list as well, but this is only my take and maybe there are cases where this would be useful
In the meantime, I am borrowing heavily from the -exclude-dirs logic to implement this, and hope to have an initial PR soon (working on the tests as we speak)

**Comment by brikis98** [comment]
> My take is that when running with include-dirs it makes no sense to add an exclusion list as well, but this is only my take and maybe there are cases where this would be useful

Typically, whitelist/blacklist logic works as follows:

* If neither is specified, include all
* If blacklist is specified, include all except the items in the blacklist
* If whitelist is specified, only include items in the whitelist
* If both whitelist and blacklist are specified, include items in the whitelist, minus those in the blacklist

**Comment by mattiarossi** [comment]
Ok, then the flow in configstack/module.go should be:

	```
       // this will flag all modules not in the included-dirs glob as Excluded
        includedModules, err := flagIncludedDirs(crossLinkedModules, terragruntOptions)
	if err != nil {
		return []*TerraformModule{}, err
	}
	finalModules, err := flagExcludedDirs(excludedModules, terragruntOptions)
	if err != nil {
		return []*TerraformModule{}, err
	}

```

.. and a test for checking that the include/exclude thing works would be to run it, say, with
`	
opts.IncludeDirs = []string{canonical(t, "../test/fixture-modules/module-c"),canonical(t, "../test/fixture-modules/module-b")}
opts.ExcludeDirs = []string{canonical(t, "../test/fixture-modules/module-b")}
`

and expecting only module-c and module-a (because of dependencies) to be flagged as included

**Comment by conorgil** [comment]
I was looking through old issues to see if anything could be closed out.

@mattiarossi It looks like your PR adding this functionality was merged. Can this issue be closed?

**Comment by brikis98** [comment]
This was indeed resolved in #626, so closing.
